### PR TITLE
Factor propagated version handling out of the sync controller

### PR DESCRIPTION
**Pull request by marun**
`PropagatedVersion` handling was previously spread all over the sync controller, and relied upon the API and an informer to manage state.  This PR extracts that code into a new `PropagatedVersionManager` type that encapsulates version management to minimize the knowledge required by the sync controller.  The version manager manages runtime state in-memory, and caches to the API to minimize updates to target clusters across sync controller restart. 

This work is related to #60, and inspired by @danehans.  Federation of cluster-scoped types will require support for cluster-scoped PropagatedVersion, and the first step towards that goal is allowing the implementation of version management to vary.

### CODE REVIEW

**Comment by marun** [comment]
@font The error handling most of your comments refer to is to be resilient in the face of API changes by something other than the version manager (e.g. errant controller or user) after the initial load.  Ideally this shouldn't happen, but there is no way to ensure absolutely that it won't.

File: pkg/controller/sync/version.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+	glog.V(4).Infof("Version manager for %q synced", targetKind)
+}
+
+func (m *propagatedVersionManager) HasSynced() bool {
+	m.RLock()
+	defer m.RUnlock()
+	return m.hasSynced
+}
+
+func (m *propagatedVersionManager) Get(template, override *unstructured.Unstructured) map[string]string {
```
[the remark this answers is not in the document]
Comment (marun) [reply pkg/controller/sync/version.go]: The point of the version manager is to hide the details of version management from the sync controller.  All the sync controller knows is that it is retrieving a map of cluster names to versions.  How it is backed or serialized is not intentionally exposed, and the naming reflects that.

File: pkg/controller/sync/version.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+		Name:      m.versionName(template.GetName()),
+	}.String()
+
+	version, ok := m.getVersion(key)
+	if !ok {
+		version := newVersion(clusterVersions, template, m.typeConfig.GetTarget().Kind, overrideVersion)
+		createdVersion, err := m.client.CoreV1alpha1().PropagatedVersions(version.Namespace).Create(version)
+		if errors.IsAlreadyExists(err) {
+			createdVersion, err = m.client.CoreV1alpha1().PropagatedVersions(version.Namespace).Get(version.Name, metav1.GetOptions{})
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (marun) [reply pkg/controller/sync/version.go]: How so?  The scope of this error is limited to the containing if block.

[the remark this answers is not in the document]
Comment (marun) [reply pkg/controller/sync/version.go]: Ah, good point.

I have the follow-on to this (#276) in-progress, and I've simplified the [method](https://github.com/kubernetes-sigs/federation-v2/pull/276/files#diff-c6e20f7b6611852ce797fb15b4fa5a34R312) that writes to the API.  Do you think it makes sense to try to get this PR merged in roughly its current shape or would it be better to collapse #276 into this review?

[the remark this answers is not in the document]
Comment (font) [reply pkg/controller/sync/version.go]: No need to collapse as we can merge this in its current shape.

**Pull request merged**